File: android/app/src/main/java/com/calc/emulator/EmulatorBridge.kt
package com.calc.emulator

import android.content.Context
import android.graphics.Bitmap
import android.util.Log

/**
 * JNI bridge to the emulator core with dynamic backend support.
 * Supports switching between Rust and CEmu backends at runtime.
 */
class EmulatorBridge {
    companion object {
        private const val TAG = "EmulatorBridge"

        // Singleton-like initialization tracking
        private var initialized = false

        // Track which backend libraries are available
        private val loadedBackends = mutableListOf<String>()

        init {
            // Load JNI bridge first
            try {
                System.loadLibrary("emu_jni")
                Log.i(TAG, "JNI loader library loaded successfully")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load JNI loader library", e)
            }

            // Preload backend libraries via System.loadLibrary to ensure they're
            // packaged in the APK. dlopen() will then find them in the native lib dir.
            tryLoadBackend("rust")
            tryLoadBackend("cemu")
        }

        private fun tryLoadBackend(name: String) {
            try {
                System.loadLibrary("emu_$name")
                loadedBackends.add(name)
                Log.i(TAG, "Backend library loaded: emu_$name")
            } catch (e: UnsatisfiedLinkError) {
                Log.d(TAG, "Backend library not available: emu_$name")
            }
        }

        /**
         * Initialize the native library with the native library directory.
         * Must be called once before any other operations.
         */
        fun initialize(context: Context) {
            if (!initialized) {
                val nativeLibDir = context.applicationInfo.nativeLibraryDir
                val cacheDir = context.cacheDir.absolutePath
                nativeInit(nativeLibDir, cacheDir)
                initialized = true
                Log.i(TAG, "Initialized with native lib dir: $nativeLibDir, cache dir: $cacheDir")
            }
        }

        /**
         * Get list of available backend names.
         * @return List of backend names (e.g., ["rust", "cemu"])
         */
        fun getAvailableBackends(): List<String> {
            return nativeGetAvailableBackends()?.toList() ?: emptyList()
        }

        /**
         * Check if multiple backends are available (show settings toggle).
         */
        fun hasMultipleBackends(): Boolean {
            return getAvailableBackends().size > 1
        }

        // Static native methods
        @JvmStatic
        private external fun nativeInit(nativeLibDir: String, cacheDir: String)

        @JvmStatic
        private external fun nativeGetAvailableBackends(): Array<String>?
    }

    // Native handle (pointer to Emu struct)
    private var handle: Long = 0

    // Cached framebuffer dimensions
    private var width: Int = 320
    private var height: Int = 240

    // Pixel buffer for framebuffer transfer
    private var pixelBuffer: IntArray = IntArray(width * height)

    /**
     * Get the currently active backend name.
     * @return Backend name or null if none loaded
     */
    fun getCurrentBackend(): String? {
        return nativeGetCurrentBackend()
    }

    /**
     * Switch to a different backend.
     * This will destroy the current emulator instance if one exists.
     * After calling this, you must call create() again.
     *
     * @param backendName Name of the backend (e.g., "rust" or "cemu")
     * @return true if successful
     */
    fun setBackend(backendName: String): Boolean {
        if (handle != 0L) {
            Log.w(TAG, "Destroying existing emulator before backend switch")
            destroy()
        }

        val success = nativeSetBackend(backendName)
        if (success) {
            Log.i(TAG, "Backend switched to: $backendName")
        } else {
            Log.e(TAG, "Failed to switch backend to: $backendName")
        }
        return success
    }

    /**
     * Create the emulator instance.
     * Must be called before any other methods.
     * @return true if successful
     */
    fun create(): Boolean {
        if (handle != 0L) {
            Log.w(TAG, "Emulator already created")
            return true
        }

        handle = nativeCreate()
        if (handle == 0L) {
            Log.e(TAG, "Failed to create emulator")
            return false
        }

        // Get framebuffer dimensions
        width = nativeGetWidth(handle)
        height = nativeGetHeight(handle)
        pixelBuffer = IntArray(width * height)

        Log.i(TAG, "Emulator created: ${width}x${height} (backend: ${getCurrentBackend()})")
        return true
    }

    /**
     * Destroy the emulator instance.
     * Must be called when done to free resources.
     */
    fun destroy() {
        if (handle != 0L) {
            nativeDestroy(handle)
            handle = 0
            Log.i(TAG, "Emulator destroyed")
        }
    }

    /**
     * Load ROM data into the emulator.
     * @param romBytes ROM file contents
     * @return 0 on success, negative error code on failure
     */
    fun loadRom(romBytes: ByteArray): Int {
        if (handle == 0L) {
            Log.e(TAG, "loadRom: emulator not created")
            return -1
        }
        return nativeLoadRom(handle, romBytes)
    }

    /**
     * Reset the emulator to initial state.
     */
    fun reset() {
        if (handle != 0L) {
            nativeReset(handle)
        }
    }

    /**
     * Power on the emulator (simulate ON key press+release).
     * Must be called after loadRom() to start execution.
     */
    fun powerOn() {
        if (handle != 0L) {
            nativePowerOn(handle)
        }
    }

    /**
     * Run emulation for the specified number of cycles.
     * @param cycles Number of cycles to execute
     * @return Number of cycles actually executed
     */
    fun runCycles(cycles: Int): Int {
        if (handle == 0L) return 0
        return nativeRunCycles(handle, cycles)
    }

    /**
     * Get the framebuffer width.
     */
    fun getWidth(): Int = width

    /**
     * Get the framebuffer height.
     */
    fun getHeight(): Int = height

    /**
     * Copy the current framebuffer to a bitmap.
     * @param bitmap Target bitmap (must be width x height, ARGB_8888)
     * @return true on success
     */
    fun copyFramebufferToBitmap(bitmap: Bitmap): Boolean {
        if (handle == 0L) return false

        val result = nativeCopyFramebuffer(handle, pixelBuffer)
        if (result != 0) {
            Log.e(TAG, "Failed to copy framebuffer: $result")
            return false
        }

        bitmap.setPixels(pixelBuffer, 0, width, 0, 0, width, height)
        return true
    }

    /**
     * Set key state.
     * @param row Key row (0-7)
     * @param col Key column (0-7)
     * @param down true if pressed, false if released
     */
    fun setKey(row: Int, col: Int, down: Boolean) {
        if (handle != 0L) {
            nativeSetKey(handle, row, col, down)
        }
    }

    /**
     * Check if emulator is created.
     */
    fun isCreated(): Boolean = handle != 0L

    /**
     * Enable instruction trace for debugging.
     * @param count Number of instructions to trace
     */
    fun enableInstTrace(count: Int) {
        // Debug stub - tracing not yet implemented in JNI
        Log.d(TAG, "enableInstTrace($count) - stub")
    }

    /**
     * Arm instruction trace to start on next wake from HALT.
     * @param count Number of instructions to trace after wake
     */
    fun armInstTraceOnWake(count: Int) {
        // Debug stub - tracing not yet implemented in JNI
        Log.d(TAG, "armInstTraceOnWake($count) - stub")
    }

    /**
     * Drain pending emulator log lines (if any).
     */
    fun drainLogs(): List<String> {
        if (handle == 0L) return emptyList()
        return nativeDrainLogs(handle)?.toList().orEmpty()
    }

    /**
     * Get the backlight brightness level (0-255).
     * Returns 0 when backlight is off (screen should be black).
     */
    fun getBacklight(): Int {
        if (handle == 0L) return 0
        return nativeGetBacklight(handle)
    }

    /**
     * Check if LCD is on (should display content).
     * Returns true when LCD should show content, false when LCD is off (show black).
     * This matches CEmu's "LCD OFF" detection.
     */
    fun isLcdOn(): Boolean {
        if (handle == 0L) return false
        return nativeIsLcdOn(handle)
    }

    // MARK: - State Persistence

    /**
     * Get the size required for save state buffer.
     * @return Size in bytes, or 0 if not available
     */
    fun saveStateSize(): Long {
        if (handle == 0L) return 0
        return nativeSaveStateSize(handle)
    }

    /**
     * Save the current emulator state.
     * @return State data as ByteArray, or null on failure
     */
    fun saveState(): ByteArray? {
        if (handle == 0L) return null

        val size = nativeSaveStateSize(handle)
        if (size <= 0) return null

        val buffer = ByteArray(size.toInt())
        val result = nativeSaveState(handle, buffer)

        return if (result >= 0) buffer else null
    }

    /**
     * Load a saved emulator state.
     * @param stateData Previously saved state data
     * @return 0 on success, negative error code on failure
     */
    fun loadState(stateData: ByteArray): Int {
        if (handle == 0L) return -1
        return nativeLoadState(handle, stateData)
    }

    // Instance native methods
    private external fun nativeGetCurrentBackend(): String?
    private external fun nativeSetBackend(backendName: String): Boolean
    private external fun nativeCreate(): Long
    private external fun nativeDestroy(handle: Long)
    private external fun nativeLoadRom(handle: Long, romBytes: ByteArray): Int
    private external fun nativeReset(handle: Long)
    private external fun nativePowerOn(handle: Long)
    private external fun nativeRunCycles(handle: Long, cycles: Int): Int
    private external fun nativeGetWidth(handle: Long): Int
    private external fun nativeGetHeight(handle: Long): Int
    private external fun nativeCopyFramebuffer(handle: Long, outArgb: IntArray): Int
    private external fun nativeSetKey(handle: Long, row: Int, col: Int, down: Boolean)
    private external fun nativeDrainLogs(handle: Long): Array<String>?
    private external fun nativeGetBacklight(handle: Long): Int
    private external fun nativeIsLcdOn(handle: Long): Boolean
    private external fun nativeSaveStateSize(handle: Long): Long
    private external fun nativeSaveState(handle: Long, outData: ByteArray): Int
    private external fun nativeLoadState(handle: Long, stateData: ByteArray): Int
}

File: ios/Calc/Bridge/EmulatorBridge.swift
//
//  EmulatorBridge.swift
//  Calc
//
//  Swift wrapper for the emulator core C API.
//  Provides thread-safe access to emulator functions.
//  Supports runtime backend switching when multiple backends are available.
//

import Foundation
import CoreGraphics
import os.log

/// Bridge to the native emulator core.
/// Thread-safe wrapper around the C API defined in emu_backend.h.
/// Supports runtime switching between Rust and CEmu backends when both are available.
class EmulatorBridge {
    private static let logger = Logger(subsystem: "com.calc.emulator", category: "EmulatorBridge")

    /// Native emulator handle (opaque pointer to Emu struct)
    private var handle: OpaquePointer?

    /// Lock for thread-safe access to emulator
    private let lock = NSLock()

    /// Cached framebuffer dimensions
    private(set) var width: Int32 = 320
    private(set) var height: Int32 = 240

    /// Log buffer for messages from emulator
    private var logBuffer: [String] = []
    private let logLock = NSLock()
    private static let maxLogs = 200

    /// Static log callback for C interop
    private static var sharedInstance: EmulatorBridge?

    init() {
        EmulatorBridge.sharedInstance = self
    }

    deinit {
        destroy()
        if EmulatorBridge.sharedInstance === self {
            EmulatorBridge.sharedInstance = nil
        }
    }

    // MARK: - Lifecycle

    /// Create the emulator instance.
    /// Must be called before any other methods.
    /// - Returns: true if successful
    func create() -> Bool {
        lock.lock()
        defer { lock.unlock() }

        if handle != nil {
            Self.logger.warning("Emulator already created")
            return true
        }

        // Set up log callback
        emu_set_log_callback { message in
            guard let message = message else { return }
            let str = String(cString: message)
            EmulatorBridge.sharedInstance?.appendLog(str)
        }

        handle = emu_create()

        if handle == nil {
            Self.logger.error("Failed to create emulator instance")
            return false
        }

        // Get framebuffer dimensions
        var w: Int32 = 0
        var h: Int32 = 0
        _ = emu_framebuffer(handle, &w, &h)
        width = w > 0 ? w : 320
        height = h > 0 ? h : 240

        Self.logger.info("Emulator created: \(self.width)x\(self.height)")
        return true
    }

    /// Destroy the emulator instance.
    /// Must be called when done to free resources.
    func destroy() {
        lock.lock()
        defer { lock.unlock() }

        if let h = handle {
            emu_destroy(h)
            handle = nil
            Self.logger.info("Emulator destroyed")
        }
    }

    /// Check if emulator is created.
    var isCreated: Bool {
        lock.lock()
        defer { lock.unlock() }
        return handle != nil
    }

    // MARK: - ROM Loading

    /// Load ROM data into the emulator.
    /// - Parameter data: ROM file contents
    /// - Returns: 0 on success, negative error code on failure
    func loadRom(_ data: Data) -> Int32 {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else {
            Self.logger.error("loadRom: emulator not created")
            return -1
        }

        Self.logger.info("Loading ROM: \(data.count) bytes")

        let result = data.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) -> Int32 in
            guard let ptr = bytes.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -2
            }
            return emu_load_rom(h, ptr, data.count)
        }

        if result != 0 {
            Self.logger.error("loadRom: emu_load_rom returned \(result)")
        }

        return result
    }

    /// Reset the emulator to initial state.
    func reset() {
        lock.lock()
        defer { lock.unlock() }

        if let h = handle {
            Self.logger.info("Resetting emulator")
            emu_reset(h)
        }
    }

    /// Power on the emulator (simulate ON key press+release to wake from reset).
    /// Must be called after loadRom() to start execution.
    func powerOn() {
        lock.lock()
        defer { lock.unlock() }

        if let h = handle {
            Self.logger.info("Powering on emulator")
            emu_power_on(h)
        }
    }

    // MARK: - Execution

    /// Run emulation for the specified number of cycles.
    /// - Parameter cycles: Number of cycles to execute
    /// - Returns: Number of cycles actually executed
    func runCycles(_ cycles: Int32) -> Int32 {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return 0 }
        return emu_run_cycles(h, cycles)
    }

    // MARK: - Framebuffer

    /// Get the current framebuffer as ARGB8888 pixels.
    /// - Returns: Tuple of (pointer to pixel data, width, height), or nil if unavailable
    func framebuffer() -> (UnsafePointer<UInt32>?, width: Int32, height: Int32) {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else {
            return (nil, 0, 0)
        }

        var w: Int32 = 0
        var h32: Int32 = 0
        let fb = emu_framebuffer(h, &w, &h32)

        return (fb, w, h32)
    }

    /// Create a CGImage from the current framebuffer.
    /// - Returns: CGImage of the screen, or nil if unavailable
    func makeImage() -> CGImage? {
        let (fb, w, h) = framebuffer()

        guard let fb = fb, w > 0, h > 0 else {
            return nil
        }

        let colorSpace = CGColorSpaceCreateDeviceRGB()

        // ARGB8888 format: alpha in high byte, then R, G, B
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue)

        guard let context = CGContext(
            data: UnsafeMutableRawPointer(mutating: fb),
            width: Int(w),
            height: Int(h),
            bitsPerComponent: 8,
            bytesPerRow: Int(w) * 4,
            space: colorSpace,
            bitmapInfo: bitmapInfo.rawValue
        ) else {
            return nil
        }

        return context.makeImage()
    }

    // MARK: - Input

    /// Set key state.
    /// - Parameters:
    ///   - row: Key row (0-7)
    ///   - col: Key column (0-7)
    ///   - down: true if pressed, false if released
    func setKey(row: Int32, col: Int32, down: Bool) {
        lock.lock()
        defer { lock.unlock() }

        if let h = handle {
            Self.logger.debug("setKey: row=\(row) col=\(col) down=\(down)")
            emu_set_key(h, row, col, down ? 1 : 0)
        }
    }

    // MARK: - LCD State

    /// Check if LCD is on (should display content).
    /// - Returns: true when LCD should show content, false when LCD is off (show black)
    func isLcdOn() -> Bool {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return false }
        return emu_is_lcd_on(h) != 0
    }

    /// Get the backlight brightness level.
    /// - Returns: 0-255, where 0 = off (screen black)
    func getBacklight() -> UInt8 {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return 0 }
        return emu_get_backlight(h)
    }

    // MARK: - Save State

    /// Get the size required for save state buffer.
    /// - Returns: Size in bytes, or 0 if not available
    func saveStateSize() -> Int {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return 0 }
        return emu_save_state_size(h)
    }

    /// Save the current emulator state.
    /// - Returns: State data, or nil on failure
    func saveState() -> Data? {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return nil }

        let size = emu_save_state_size(h)
        guard size > 0 else { return nil }

        var buffer = Data(count: size)
        let result = buffer.withUnsafeMutableBytes { (bytes: UnsafeMutableRawBufferPointer) -> Int32 in
            guard let ptr = bytes.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -1
            }
            return emu_save_state(h, ptr, size)
        }

        return result >= 0 ? buffer : nil
    }

    /// Load a saved emulator state.
    /// - Parameter data: Previously saved state data
    /// - Returns: 0 on success, negative error code on failure
    func loadState(_ data: Data) -> Int32 {
        lock.lock()
        defer { lock.unlock() }

        guard let h = handle else { return -1 }

        return data.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) -> Int32 in
            guard let ptr = bytes.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -2
            }
            return emu_load_state(h, ptr, data.count)
        }
    }

    // MARK: - Backend Management

    /// Get the list of available backends.
    /// - Returns: Array of backend names (e.g., ["rust", "cemu"])
    static func getAvailableBackends() -> [String] {
        guard let cString = emu_backend_get_available() else {
            return []
        }
        let available = String(cString: cString)
        if available.isEmpty {
            return []
        }
        return available.components(separatedBy: ",")
    }

    /// Get the number of available backends.
    /// - Returns: Number of backends (0, 1, or 2)
    static func getBackendCount() -> Int {
        return Int(emu_backend_count())
    }

    /// Get the current backend name.
    /// - Returns: Current backend name, or nil if no backend is loaded
    static func getCurrentBackend() -> String? {
        guard let cString = emu_backend_get_current() else {
            return nil
        }
        return String(cString: cString)
    }

    /// Set the active backend.
    /// Note: This must be called before creating an emulator instance,
    /// or after destroying the current instance.
    /// - Parameter name: Backend name ("rust" or "cemu")
    /// - Returns: true if successful
    static func setBackend(_ name: String) -> Bool {
        let result = emu_backend_set(name)
        if result == 0 {
            logger.info("Backend switched to: \(name)")
            return true
        } else {
            logger.error("Failed to switch backend to: \(name)")
            return false
        }
    }

    /// Check if backend switching is available (multiple backends linked).
    /// - Returns: true if more than one backend is available
    static func isBackendSwitchingAvailable() -> Bool {
        return getBackendCount() > 1
    }

    // MARK: - Logging

    private func appendLog(_ message: String) {
        logLock.lock()
        defer { logLock.unlock() }

        logBuffer.append(message)
        if logBuffer.count > Self.maxLogs {
            logBuffer.removeFirst(logBuffer.count - Self.maxLogs)
        }
    }

    /// Drain pending log messages.
    /// - Returns: Array of log messages since last drain
    func drainLogs() -> [String] {
        logLock.lock()
        defer { logLock.unlock() }

        let logs = logBuffer
        logBuffer.removeAll()
        return logs
    }
}
